src/extractor/myexperience/index.ts

```diff
@@ -0,0 +1,426 @@
+import puppeteer, { Browser, ElementHandle, Page } from "puppeteer";
+
+import {
+  getPostingListData,
+  getPostingTables,
+  getPostingTags,
+  navigateToPostingSubPage,
+} from "../common";
+import {
+  delay,
+  getInnerText,
+  getInnerTextFallback,
+  priorityMatch,
+  splitFirst,
+} from "../util";
+import Posting, { PostingError, parsePostingData } from "./posting";
+import { parseWorkTerm } from "./posting/workTerm";
+
+type PostingCommon = Pick<Posting, "id" | "title" | "subtitle"> & {
+  statusData: Posting["status"]["data"];
+};
+type MultipleExtractOptions = {
+  maxPages?: number;
+  maxConcurrency?: number;
+  delay?: number;
+  startOnCurrent?: boolean;
+};
+
+export * from "./posting";
+export const POSTING_URL =
+  "https://myexperience.sfu.ca/myAccount/co-op/postings.htm";
+export const DASHBOARD_URL =
+  "https://myexperience.sfu.ca/myAccount/dashboard.htm";
+
+const POSTING_VIEWPORT = { height: 960, width: 640 };
+
+/**
+ * An extractor for job postings on WaterlooWorks
+ */
+export default class Extractor {
+  public browser: Browser;
+
+  public constructor(browser: Browser) {
+    this.browser = browser;
+  }
+
+  /**
+   * Launches a new browser and extractor
+   * @returns A new extractor attached to a new browser
+   */
+  public static async launch() {
+    const browser = await puppeteer.launch({ headless: false });
+    return new Extractor(browser);
+  }
+
+  /**
+   * Navigate to the login page for the job board.
+   * This function will wait indefinitely until the user has logged in.
+   *
+   * @param keepPage - Whether to keep the login page open
+   * @returns The login page if `keepPage` is set, otherwise `undefined`
+   */
+  public async login<R extends boolean = false>(
+    keepPage?: R
+  ): Promise<R extends true ? Page : undefined> {
+    const page = await this.browser.newPage();
+    await page.goto(DASHBOARD_URL, { waitUntil: "domcontentloaded" });
+    await page.waitForSelector(
+      'a.button[href*="/home.htm"], a[href="/logout.htm"]',
+      {timeout: 30000} // longer await timeout becuase of MFA
+    );
+    if (!(await page.$('a[href="/logout.htm"]'))) {
+      await page.click('a.button[href*="/home.htm"]');
+      await page.waitForSelector(
+        'a[href*="waterloo.htm"][href*="action=login"]',
+        {
+          visible: true,
+        }
+      );
+      await page.click('a[href*="waterloo.htm"][href*="action=login"]');
```

Comment:
This block was probably not tested, since it still references stuff from Waterloo.

Reply:
Forgot to ask what that part does

Reply:
Conditional on line 72 checks if we are already logged in (if navigating to the dashboard url results in a button to logout, then we know we are already logged in). Otherwise, it enters line 73 and clicks a few elements to bring the page to the login screen. We wait there indefinitely (line 87) since the user needs to interact with the page.